src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp

```diff
@@ -246,116 +235,363 @@ class Multiphase :
+                                                         strainIncrement,
+                                                         stack.totalStress,
+                                                         stack.dTotalStress_dPressure,
+                                                         stack.dTotalStress_dTemperature,
+                                                         stack.stiffness,
+                                                         porosity,
+                                                         porosity_n,
+                                                         dPorosity_dVolStrain,
+                                                         dPorosity_dPressure,
+                                                         dPorosity_dTemperature,
+                                                         dSolidDensity_dPressure );
+
+    // Step 2: compute the body force
+    if( m_gravityAcceleration > 0.0 )
     {
-      localIndex const localNodeIndex = m_elemsToNodes( k, a );
+      computeBodyForce( k, q,
+                        porosity,
+                        dPorosity_dVolStrain,
+                        dPorosity_dPressure,
+                        dPorosity_dTemperature,
+                        dSolidDensity_dPressure,
+                        stack );
+    }
+
+    // Step 3: compute fluid mass increment
+    computeFluidIncrement( k, q,
+                           porosity,
+                           porosity_n,
+                           dPorosity_dVolStrain,
+                           dPorosity_dPressure,
+                           dPorosity_dTemperature,
+                           stack );
+
+    // Step 4: compute pore volume constraint
+    computePoreVolumeConstraint( k,
+                                 porosity,
+                                 dPorosity_dVolStrain,
+                                 dPorosity_dPressure,
+                                 dPorosity_dTemperature,
+                                 stack );
+  }
+
+  /**
+   * @brief Helper function to compute the body force term and its derivatives wrt primary variables
+   * @tparam FUNC the type of the function that can be used to customize the kernel
+   * @param[in] k the element index
+   * @param[in] q the quadrature point index
+   * @param[in] porosity the element porosity
+   * @param[in] dPorosity_dVolStrain the derivative of porosity wrt volumetric strain increment
+   * @param[in] dPorosity_dPressure the derivative of porosity wrt pressure
+   * @param[in] dPorosity_dTemperature the derivative of porosity wrt temperature
+   * @param[in] dSolidDensity_dPressure the derivative of solid density wrt pressure
+   * @param[inout] stack the stack variables
+   * @param[inout] bodyForceKernelOp the lambda function to customize the kernel
+   */
+  template< typename FUNC = NoOpFunc >
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  void
+  computeBodyForce( localIndex const k,
```

Comment:
What's exactly a `bodyForce`?

Reply:
Done, I added a comment to explain what the body force is (standard name of `\rho g` in the linear momentum balance equation)